dask/delayed.py

```diff
@@ -118,7 +118,25 @@ def unpack_collections(expr):
                 if hasattr(expr, f.name)  # if init=False, field might not exist
             ]
         )
-
+        if not collections:
+            return expr, ()
+        try:
+            _fields = {
+                f.name: getattr(expr, f.name)
+                for f in fields(expr)
+                if hasattr(expr, f.name)
+            }
+            replace(expr, **_fields)
+        except TypeError as e:
+            raise TypeError(
+                f"Failed to unpack {typ} instance. "
+                "Note that using a custom __init__ is not supported."
+            ) from e
+        except ValueError as e:
+            raise ValueError(
+                f"Failed to unpack {typ} instance. "
+                "Note that using fields with `init=False` are not supported."
+            ) from e
```

Comment:
Unfortunately, there is another (slightly different) version of `unpack_collections` in `dask.base`. We should probably give the same treatment there.

I can trigger the same bug with
```python
import dask
from dataclasses import dataclass

@dataclass
class MyDataClass:
    arg1: int
    def __init__(self, arg):
        self.arg1 = arg
        
dask.compute([dask.delayed(1), MyDataClass(1)])
```

Reply:
I wasn't aware of that, thanks! I'll adjust and test it accordingly.

Reply:
Turns out that the implementation of `dask.base.unpack_collections` makes this not trivial. I think we can move forward with this PR without that fix. @hendrikmakait, would you mind opening a new issue as a follow-up to this one?

Reply:
Follow-up issue: #9370